keps/78-dynamically-reclaiming-resources/README.md

```diff
@@ -0,0 +1,169 @@
+  - [Workload](#workload)
+    - [API](#api-1)
+    - [<code>pkg/workload</code>](#)
+    - [Controller](#controller)
+  - [Jobframwork](#jobframwork)
+  - [Batch/Job](#batchjob)
+- [Testing Plan](#testing-plan)
+  - [NonRegression](#nonregression)
+  - [Unit Tests](#unit-tests)
+  - [Integration tests](#integration-tests)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+This proposal allows Kueue to reclaim the resources of a successfully completed Pod of a Workload even before the whole Workload completes its execution. Freeing the unused resources of a Workload earlier may allow more pending Workloads to be admitted. This will allow Kueue to utilize the existing resources of the cluster much more efficiently.
+
+In the remaining of this document:
+1. *Job* refers to any kind job supported by Kueue including `batch/job`, `MPIJob`, `RayJob`, etc.
+1. *Pods of a Workload* means the same as *Pods of a Job* and vice-versa.
+
+
+## Motivation
+
+Currently, the assigned to a Job are reclaimed by Kueue only when the whole Job finishes.
+For Jobs with multiple Pods, the resources will be reclaimed only after the last Pod of the Job finishes.
+This is not efficient as the Jobs might have different needs during execution, for instance the needs of a `batch/job` having the `parallelism` equal to `completions` will decrease with every Pod finishing it's execution.
+
+### Goals
+
+- Utilize the unused resources of the successfully completed Pods of a running Job.
+
+### Non-Goals
+
+- Preempting a Workload or Pods of a Workload to free resources for a pending Workload.
+- Partially admitting a Workload.
+- Monitor the Job's pod execution.
+
+## Proposal
+
+Reclaiming the resources of the succeeded Pods of a running Job as soon as the Pod completes its execution.
+
+We propose to add a new field `.status.ReclaimablePods` to the Workload API. `.status.ReclaimablePods` is a list that holds the count of Pods belonging to a PodSet whose resources are no longer needed and could be reclaimed by Kueue.
+
+### Pod Completion Accounting
+
+Since the ability of actively monitoring pods execution of a job is not in the scope of the core Kueue implementation, the number of pods for which the resources are no longer needed for each PodSet should be reported by each framework specific `GenericJob` implementation.
+
+For this purpose the `GenericJob` interface should be changed to either:
+
+1. Add an additional method able to report this
+
+```go
+type GenericJob interface {
+    // ...
+
+   // Get reclaimable pods.
+   ReclaimablePods() map[string]int
+}
+
+```
+
+2. Modify the current `Finished` method
+
+```go
+type GenericJob interface {
+   // ...
+
+   // Get reclaimable pods.
+   Finished() (condition metav1.Condition, reclaimablePods map[string]int)
+}
+
+```
+
+#### Example (`batch/job`)
+
+Having a job defined with **P** `parallelism` and **C** `completions`, and **n** number of completed pod executions,
```

Comment:
Does the `completions` mean `.status.succeeded`?

Reply:
It would be good to clarify conditions to handle as `completions`.

Reply:
Thank you for pointing out that!

Reply:
For `batch/Job` yes, for  workloads we only need to know the number of pods we no longer need to reserve quota for.

Reply:
Not resolved yet.

Reply:
I can't say I see a very strong connection between them and this KEP, #kubernetes/enhancements/pull/3940 will improve the accuracy of `.status.succeeded` but it has no direct impact on on the current issue.  With #kubernetes/enhancements/pull/3967 we might go from running with the full umber of pods to `Finished` but also it has no direct impact. Am I missing something?

I can add a new section with related issues and list them there, wold that be OK?

Reply:
done